File: 0_1.baekjoon/dp/java/src/BOJ20542.java
import java.io.*;
import java.util.*;
public class BOJ20542 {
    public static void main(String args[]) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        int[][] dp = new int[n + 1][m + 1];

        for (int i=0;i<=n;i++) {
            dp[i][0] = i;
        }

        for (int i=0;i<=m;i++) {
            dp[0][i] = i;
        }

        char[] result = br.readLine().toCharArray();
        char[] solution = br.readLine().toCharArray();

        for (int i=1;i<=n;i++) {
            for (int j=1;j<=m;j++) {
                if (isMatched(result[i - 1], solution[j - 1])) {
                    dp[i][j] = dp[i - 1][j - 1];
                } else {
                    dp[i][j] = Math.min(dp[i - 1][j], Math.min(dp[i - 1][j - 1], dp[i][j - 1])) + 1;
                }
            }
        }

        System.out.println(dp[n][m]);
    }

    private static boolean isMatched(char input, char target) {
        return input == target || (input == 'i' && (target == 'j' || target == 'l')) || (input == 'v' && target == 'w');
    }

}

File: 0_1.baekjoon/dp/kotlin/src/BOJ20542.kt
import kotlin.math.min

fun main() = with(System.`in`.bufferedReader()) {
    val (n, m) = readLine().split(" ").map { it.toInt() }
    val dp = Array(n + 1) { IntArray(m + 1)}
    val answer = readLine()
    val result = readLine()

    for (i in 1..n) {
        dp[i][0] = i
    }

    for (i in 1..m) {
        dp[0][i] = i
    }

    for (i in 1..n) {
        for (j in 1..m) {
            if (correspond(answer[i - 1], result[j - 1])) {
                dp[i][j] = dp[i - 1][j - 1]
            } else {
                dp[i][j] = min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]) + 1
            }
        }
    }

    print(dp[n][m])
}

fun correspond(input: Char, target: Char): Boolean {
    return input == target || (input == 'i' && (target == 'j' || target == 'l') || (input == 'v' && target == 'w'))
}

private fun min(i: Int, j: Int, k: Int): Int {
    return min(i, min(j, k))
}
